File: 748_shortest_completing_word/shortest_completing_word.cpp
/*
 * @Date: 2021-12-10 09:28:52
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-10 09:36:58
 */

#include <array>
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  string shortestCompletingWord(string licensePlate, vector<string> words) {
    array<int, 26> cnt{};
    for (char ch : licensePlate)
      if (isalpha(ch)) ++cnt[tolower(ch) - 'a'];
    int idx = -1;
    int words_size = words.size();
    for (int i = 0; i < words_size; ++i) {
      array<int, 26> c{};
      for (char ch : words[i]) {
        ++c[ch - 'a'];
      }
      bool ok = true;
      for (int j = 0; j < 26; ++j) {
        if (c[j] < cnt[j]) {
          ok = false;
          break;
        }
      }
      if (ok && (idx < 0 || words[i].length() < words[idx].length())) idx = i;
    }
    return words[idx];
  }
};

int main() {
  assert(Solution().shortestCompletingWord(
             "1s3 PSt", vector<string>{"step", "steps", "stripe", "stepple"}) ==
         "steps");
  assert(Solution().shortestCompletingWord(
             "1s3 456", vector<string>{"looks", "pest", "stew", "show"}) ==
         "pest");
  assert(Solution().shortestCompletingWord(
             "Ah71752", vector<string>{"suggest", "letter", "of", "husband",
                                       "easy", "education", "drug", "prevent",
                                       "writer", "old"}) == "husband");
  assert(Solution().shortestCompletingWord(
             "OgEu755",
             vector<string>{"enough", "these", "play", "wide", "wonder", "box",
                            "arrive", "money", "tax", "thus"}) == "enough");
  assert(Solution().shortestCompletingWord(
             "iMSlpe4", vector<string>{"claim", "consumer", "student", "camera",
                                       "public", "never", "wonder", "simple",
                                       "thought", "use"}) == "simple");
}

File: 748_shortest_completing_word/shortest_completing_word.rs
/*
 * @Date: 2021-12-10 09:28:57
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-10 09:48:45
 */

pub fn shortest_completing_word(license_plate: String, words: Vec<String>) -> String {
    let to_state = |t: &String| {
        t.to_lowercase()
            .bytes()
            .filter(|&c| c.is_ascii_alphabetic())
            .fold(vec![0; 26], |mut s, c| {
                s[(c - b'a') as usize] += 1;
                s
            })
    };
    let lp_state = to_state(&license_plate);
    words.iter().fold(words.join(" "), |mut ans, word| {
        let word_state = to_state(word);
        if lp_state.iter().zip(&word_state).all(|(c1, c2)| c1 <= c2) && word.len() < ans.len() {
            ans = word.clone();
        }
        ans
    })
}

fn main() {
    assert_eq!(
        shortest_completing_word(
            String::from("1s3 PSt"),
            vec!["step", "steps", "stripe", "stepple"]
                .iter()
                .map(|s| s.to_string())
                .collect()
        ),
        "steps"
    );
    assert_eq!(
        shortest_completing_word(
            String::from("1s3 456"),
            vec!["looks", "pest", "stew", "show"]
                .iter()
                .map(|s| s.to_string())
                .collect()
        ),
        "pest"
    );
    assert_eq!(
        shortest_completing_word(
            String::from("Ah71752"),
            vec![
                "suggest",
                "letter",
                "of",
                "husband",
                "easy",
                "education",
                "drug",
                "prevent",
                "writer",
                "old"
            ]
            .iter()
            .map(|s| s.to_string())
            .collect()
        ),
        "husband"
    );

    assert_eq!(
        shortest_completing_word(
            String::from("OgEu755"),
            vec![
                "enough", "these", "play", "wide", "wonder", "box", "arrive", "money", "tax",
                "thus"
            ]
            .iter()
            .map(|s| s.to_string())
            .collect()
        ),
        "enough"
    );

    assert_eq!(
        shortest_completing_word(
            String::from("iMSlpe4"),
            vec![
                "claim", "consumer", "student", "camera", "public", "never", "wonder", "simple",
                "thought", "use"
            ]
            .iter()
            .map(|s| s.to_string())
            .collect()
        ),
        "simple"
    );
}
